src/transformers/models/falcon/modeling_falcon.py

```diff
@@ -99,19 +99,40 @@ def _set_cos_sin_cache(self, seq_len, device, dtype):
         self.cos_cached = self.cos_cached.type(dtype)
         self.sin_cached = self.sin_cached.type(dtype)
 
-    def cos_sin(self, seq_len: int, past_key_values_length: int, device="cpu", dtype=torch.bfloat16) -> torch.Tensor:
+    def cos_sin(
```

Comment:
I'm gonna be a bit noisy here, but this looks a LOT like the rotary embedding we have in Llama no? 
The query expansion is also supported there, not sure how much of an overhead it is to first apply rotary then expand:
```python 
        query_states, key_states = apply_rotary_pos_emb(query_states, key_states, cos, sin, position_ids)

        if past_key_value is not None:
            # reuse k, v, self_attention
            key_states = torch.cat([past_key_value[0], key_states], dim=2)
            value_states = torch.cat([past_key_value[1], value_states], dim=2)

        past_key_value = (key_states, value_states) if use_cache else None

        # repeat k/v heads if n_kv_heads < n_heads
        key_states = repeat_kv(key_states, self.num_key_value_groups)
        value_states = repeat_kv(value_states, self.num_key_value_groups)
```
and also storing the full size key and values is less memory efficient no? (unrelated to the PR).

Reply:
I *think* they are the same. So, we would benefit from copying the structure (at least in terms of complexity) 👍 

I would like to push it to the future, though, as I'm about to go on long holidays and I'd like to enable batched generation on Falcon :D